pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -315,23 +440,166 @@ cdef inline timedelta_from_spec(object number, object frac, object unit):
     n = ''.join(number) + '.' + ''.join(frac)
     return cast_from_unit(float(n), unit)
 
+
+# ----------------------------------------------------------------------
+# Timedelta ops utilities
+
+cdef bint _validate_ops_compat(other):
+    # return True if we are compat with operating
+    if _checknull_with_nat(other):
+        return True
+    elif PyDelta_Check(other) or is_timedelta64_object(other):
+        return True
+    elif is_string_object(other):
+        return True
+    elif hasattr(other, 'delta'):
+        return True
+    return False
+
+
+def _op_unary_method(func, name):
+    def f(self):
+        return Timedelta(func(self.value), unit='ns')
+    f.__name__ = name
+    return f
+
+
+def _binary_op_method_timedeltalike(op, name):
+    # define a binary operation that only works if the other argument is
+    # timedelta like or an array of timedeltalike
+    def f(self, other):
+        if hasattr(other, 'delta') and not PyDelta_Check(other):
+            # offsets.Tick
+            return op(self, other.delta)
+
+        elif other is NaT:
+            return NaT
+
+        elif is_datetime64_object(other) or PyDateTime_CheckExact(other):
+            # the PyDateTime_CheckExact case is for a datetime object that
+            # is specifically *not* a Timestamp, as the Timestamp case will be
+            # handled after `_validate_ops_compat` returns False below
+            from ..tslib import Timestamp
+            return op(self, Timestamp(other))
+            # We are implicitly requiring the canonical behavior to be
+            # defined by Timestamp methods.
+
+        elif hasattr(other, 'dtype'):
+            # nd-array like
+            if other.dtype.kind not in ['m', 'M']:
+                # raise rathering than letting numpy return wrong answer
+                return NotImplemented
+            return op(self.to_timedelta64(), other)
+
+        elif not _validate_ops_compat(other):
+            return NotImplemented
+
+        try:
+            other = Timedelta(other)
+        except ValueError:
+            # failed to parse as timedelta
+            return NotImplemented
+
+        return Timedelta(op(self.value, other.value), unit='ns')
+
+    f.__name__ = name
+    return f
+
+
 # ----------------------------------------------------------------------
 # Timedelta Construction
 
+cdef _to_py_int_float(v):
+    # Note: This used to be defined inside _timedelta_value_kwargs
```

Comment:
should prob move to util.pyx (I know it doesn't exist, but should). though maybe not useful outside of this module.

Reply:
> util.pyx (I know it doesn't exist, but should).

Separate issue: the vast majority of usage of util is for `is_foo_object`, and those functions could be made pure cython (in C-equivalent ways as in #18059), i.e. put in a file without having dependencies on src, so we wouldn't have to futz around with `include/sources` (or in my opinion `depends/pxdfiles` since cython will take care of that for us).

Reply:
i agree though I don't think its a big deal to simply include it. having too fine grain dependencies is a problem too.